### PR TITLE
Expanding our http-pipeline sample into a full test project

### PR DESCRIPTION
**Pull request by scbedd**
Addresses the .NET side of #4674

@mario-guerra and @xboxeer for FYI. I got a couple free hours this afternoon to spend on this. Is this direction helpful to you?

@mikeharder can you give this a quick look and throw me a signoff / feedback?

All this does is expand your http-pipeline sample integration to a full test project, placing the start/stop in `setup`/`teardown` functions.

I have no idea why the APIView stuff is triggering from this changeset, but it made me notice a failing install in APIView that I figured I'd attempt to resolve.

### CODE REVIEW

**Comment by scbedd** [comment]
Hey @mario-guerra is there a language you're more familiar with that I can write this in?

This is already pretty darn close to as bare-bones as one can possibly get. `SampleBaseClass` takes care of the necessary start/stop/request modification. `SampleTestImplementation` is the single test that's actually running.

I can add an index to the readme and maybe rename them, but that's about it.

**Comment by mario-guerra** [comment]
@scbedd would it be possible to put something together in Python?

**Comment by scbedd** [comment]
> @scbedd would it be possible to put something together in Python?

Though this is less useful for the mgmt team, yes, I can. Will get back to you.

**Comment by mario-guerra** [comment]
Ok, thanks Scott.

**Comment by xboxeer** [comment]
@scbedd I tested this project and I can run it successfully, both the integration project and the test project, however when I want to use the test-proxy to record an actual Azure SDK .NET call, I am not able to get it recorded. I looked through the code seems this demo is using Azure Core raw request pipeline but in our SDK we usually use the SDK provided class, like the code below
```csharp
//Register ArmClient in DI
builder.AddClient<ArmClient, ArmClientOptions>(options =>
           {
               options.Environment = ArmEnvironment.AzurePublicCloud;
               options.Diagnostics.IsLoggingContentEnabled=true;
               options.Diagnostics.IsLoggingEnabled=true;
               Environment.SetEnvironmentVariable("HTTPS_PROXY", "https://localhost:5001");
               var client = new ArmClient(new DefaultAzureCredential(
                   new DefaultAzureCredentialOptions() 
                   { VisualStudioTenantId=context.Configuration.GetValue<string>("VisualStudioTenantId") }),
                   context.Configuration.GetValue<string>("DefaultSubscriptionId"),
                   options);
               return client;                
           });

//Actual SDK call
public class ResourceGroupService
    {
        private ArmClient client;
        public ResourceGroupService(ArmClient client)
        {
            this.client = client;
        }

        public void CreateResourceGroup(string resourceGroupName)
        {
            var resourceGroups = this.client.GetDefaultSubscription().GetResourceGroups();            
            if (resourceGroups.Exists(resourceGroupName))
            {
                return;
            }          
            ResourceGroupResource resourceGroup = resourceGroups.CreateOrUpdate(Azure.WaitUntil.Completed, resourceGroupName, new ResourceGroupData(Azure.Core.AzureLocation.WestUS2)).Value;

        }
}
```
What I did is added a new line ` Environment.SetEnvironmentVariable("HTTPS_PROXY", "https://localhost:5001");` before I create the ArmClient, my assumption is that the test-proxy will receive the request, record the request, and forward the request to actual azure service, am I missing something here?

File: tools/test-proxy/sample-clients/net/http-pipeline/Azure.Sdk.Tools.TestProxy.HttpPipelineTestsExample/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+## Run the tests
+
+- Run tests in your preferred method, either through `dotnet test` or visual studio test explorer. Both work!
+
+### Recording using Visual Studio for manual debugging
+
+Unfortunately .NET test projects have a bad habit of not honoring environment variable settings from test projects. As a result of this, if the user is debugging these tests from a local visual studio instance.
+
+- Update the RECORD_MODE static in `SampleTestClass.cs` to "record".
```
Comment (xboxeer) [inline tools/test-proxy/sample-clients/net/http-pipeline/Azure.Sdk.Tools.TestProxy.HttpPipelineTestsExample/README.md:19]: no need to change the RECORD_MODE in SampleTestBaseClasee, we can directly change env var in test method code

**Comment by scbedd** [comment]
> @scbedd I tested this project and I can run it successfully, both the integration project and the test project, however when I want to use the test-proxy to record an actual Azure SDK .NET call, I am not able to get it recorded. I looked through the code seems this demo is using Azure Core raw request pipeline but in our SDK we usually use the SDK provided class, like the code below
> 
> ```cs
> //Register ArmClient in DI
> builder.AddClient<ArmClient, ArmClientOptions>(options =>
>            {
>                options.Environment = ArmEnvironment.AzurePublicCloud;
>                options.Diagnostics.IsLoggingContentEnabled=true;
>                options.Diagnostics.IsLoggingEnabled=true;
>                Environment.SetEnvironmentVariable("HTTPS_PROXY", "https://localhost:5001");
>                var client = new ArmClient(new DefaultAzureCredential(
>                    new DefaultAzureCredentialOptions() 
>                    { VisualStudioTenantId=context.Configuration.GetValue<string>("VisualStudioTenantId") }),
>                    context.Configuration.GetValue<string>("DefaultSubscriptionId"),
>                    options);
>                return client;                
>            });
> 
> //Actual SDK call
> public class ResourceGroupService
>     {
>         private ArmClient client;
>         public ResourceGroupService(ArmClient client)
>         {
>             this.client = client;
>         }
> 
>         public void CreateResourceGroup(string resourceGroupName)
>         {
>             var resourceGroups = this.client.GetDefaultSubscription().GetResourceGroups();            
>             if (resourceGroups.Exists(resourceGroupName))
>             {
>                 return;
>             }          
>             ResourceGroupResource resourceGroup = resourceGroups.CreateOrUpdate(Azure.WaitUntil.Completed, resourceGroupName, new ResourceGroupData(Azure.Core.AzureLocation.WestUS2)).Value;
> 
>         }
> }
> ```
> 
> What I did is added a new line ` Environment.SetEnvironmentVariable("HTTPS_PROXY", "https://localhost:5001");` before I create the ArmClient, my assumption is that the test-proxy will receive the request, record the request, and forward the request to actual azure service, am I missing something here?
> 
> FYI: although dotnet test project does not support setup environment, but we can add this line of code at the beginning of test method to avoid changing the static RECORD_MODE in SampleTestBaseClass.cs `Environment.SetEnvironmentVariable("RECORD_MODE", "record");`

This is intended to be an _example_ of integrating your tests with the test-proxy. The pipeline piece that modifies the request before it fires is the piece you need. I will change this to do the same thing using a base `httpClient`.

You're going to need to modify your existing test stack with the transport options. This project _intentionally_ isn't a true proxy. If it were a true proxy and handled redirected requests from `HTTP_PROXY` env setting, how would we TEST the proxy settings?

The way this thing works is NOT by traditional proxy methods, but by adjusting your requests on the fly.

If you look at the testclass, it uses the specific client that we created for the test. For your testing, you're going to want to do _something similar_ where you create the client used by your tests, and it will take the correct action.

The way the .NET data team does this is by modifying their [RecordedTestbase](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/RecordedTestBase.cs). This maintains the reference to the testproxy functionality, and takes care of all the start/stop etc.

[Here is the confidential ledger recorded tests](https://github.com/Azure/azure-sdk-for-net/blob/64d9174e4f58b4f2c8b02b459f4edf8ea002f441/sdk/confidentialledger/Azure.Security.ConfidentialLedger/tests/ConfidentialLedgerClientLiveTests.cs#L36) There is a lot of code, but should be able to follow the `InstrumentClient`. Effectively, it's a standard client that has been created with additional options to cause the URL rewrite.

Given that a lot of the mgmt packages have gone to track 2, I would expect that the URL rewrite capability is actually present via transportoptions! You just need to update how it is passed in to your test!

**Review by mikeharder — APPROVED** [review]

**Pull request merged**